codec/src/main/java/io/netty/handler/codec/compression/CompressionEncoder.java

```diff
@@ -0,0 +1,167 @@
+
+    /**
+     * Default constructor for {@link CompressionEncoder} which will try to use a direct
+     * {@link ByteBuf} as a target for the encoded messages.
+     */
+    protected CompressionEncoder(CompressionFormat format) {
+        this(format, true);
+    }
+
+    /**
+     * Creates an instance with specified preference of a target {@link ByteBuf} for the encoded messages.
+     * It uses if compression algorithm needs a byte array as a target.
+     *
+     * @param preferDirect {@code true} if a direct {@link ByteBuf} should be tried to be used as a target for
+     *                     the encoded messages. If {@code false} is used it will allocate a heap
+     *                     {@link ByteBuf}, which is backed by an byte array.
+     */
+    protected CompressionEncoder(CompressionFormat format, boolean preferDirect) {
+        super(preferDirect);
+        if (format == null) {
+            throw new NullPointerException("format");
+        }
+        this.format = format;
+    }
+
+    /**
+     * Returns a format of current compression algorithm.
+     */
+    public final CompressionFormat format() {
+        return format;
+    }
+
+    /**
+     * Returns {@code true} if and only if the end of the compressed stream
+     * has been reached.
+     */
+    public abstract boolean isClosed();
+
+    /**
+     * Close this {@link CompressionEncoder} and so finish the encoding.
+     *
+     * The returned {@link ChannelFuture} will be notified once the
+     * operation completes.
+     *
+     * @throws UnsupportedOperationException
+     *      if current encoder does not support closing
+     */
+    public ChannelFuture close() {
+        return close(ctx().newPromise());
+    }
+
+    /**
+     * Close this {@link CompressionEncoder} and so finish the encoding.
+     * The given {@link ChannelFuture} will be notified once the operation
+     * completes and will also be returned.
+     *
+     * @throws UnsupportedOperationException
+     *      if current encoder does not support closing
+     */
+    public ChannelFuture close(final ChannelPromise promise) {
+        final ChannelHandlerContext ctx = ctx();
+        EventExecutor executor = ctx.executor();
+        if (executor.inEventLoop()) {
+            return finishEncode(ctx, promise);
+        } else {
+            final ChannelPromise p = ctx.newPromise();
+            executor.execute(new OneTimeTask() {
+                @Override
+                public void run() {
+                    ChannelFuture f = finishEncode(ctx(), p);
+                    f.addListener(new ChannelPromiseNotifier(promise));
+                }
+            });
+            return p;
+        }
+    }
+
+    @Override
+    public void close(final ChannelHandlerContext ctx, final ChannelPromise promise) throws Exception {
+        ChannelFuture f = finishEncode(ctx, ctx.newPromise());
+        f.addListener(new ChannelFutureListener() {
```

Comment:
only add the listener to the future if its not done. If its done you can just call `ctx.close(...)` directly and so reduce object allocations